### DISCUSSION TITLE
CC: Tweaked 1.106 for 1.19.4 and 1.20

### DISCUSSION THREAD

**Discussion by SquidDev** [cc-tweaked/CC-Tweaked#1514] [Announcements]
Hello again! It's another month, and that means (well, sometimes), it's time for another CC: Tweaked update. 

# SSRF security vulnerability
First, some serious news. The wonderful @JLLeitschuh found a serious exploit when running a Minecraft server on various popular cloud providers (AWS, Digital Ocean, Oracle). Please see [the announcement](https://github.com/cc-tweaked/CC-Tweaked/discussions/1513) and [advisory](https://github.com/cc-tweaked/CC-Tweaked/security/advisories/GHSA-7p4w-mv69-2wm2) for more information.

# Relicensing update
Earlier this year we started [relicensing (or as much as possible) under the MPL 2.0](https://github.com/cc-tweaked/CC-Tweaked/issues/1339). Since then there's been a lot of progress, with over 80 people agreeing to relicense their contributions.

There's still a little more work to be done here (there's 8 people major contributions I've not been able to contact), but this means over half the mod is now licensed under the MPL 2.0:

![A sunburst graph of the the CC:T's source code, showing what files are CCPL and MPL. The centre circle is purple (representing an even split of MPL and CCPL code), but the leafe nodes are more starkly. Older packages like turtle are incredibly red, while newer ones like command and data are blue.](https://github.com/cc-tweaked/CC-Tweaked/assets/4346137/e9a47cb6-2025-4ce5-b7e3-df58272c9f5f)

# Turtle upgrade overhaul
@SirEdvin did some incredible work this month of improving the API for turtle and pocket computer upgrades. Upgrades can now read and write NBT from the upgrade item when equipped. This unlocks a whole bunch of additional functionality, most notably tool enchantments and durability:

![A screenshot of a ComputerCraft turtle's GUI. A diamond pickaxe is equipped - it's tooltip shows it has the Silk Touch enchantment and has 1549/1561 durability.](https://github.com/cc-tweaked/CC-Tweaked/assets/4346137/15385019-1be9-4b49-a9e2-c959b4c400eb)

You may also notice that the turtle UI now contains additional slots for upgrades. You can now easily change upgrades directly, without having to use the `equip` command.

Note that these new features are _not_ enabled by default right now. Instead, you can use a datapack (such as [netherite-tools](https://github.com/cc-tweaked/netherite-tools)) to enable this functionality.

# Glow in the dark computers
This is very silly, but I love it so. While monitors have glowed in the dark for years, computers now do the same:

![A screenshot of a ComputerCraft computer in a very dark room. One can barely make anything out, aside from the computer's cursor which glows a bright white.](https://github.com/cc-tweaked/CC-Tweaked/assets/4346137/7bb952ed-240e-4f54-9c1a-ad54e1eefe5f)

I'm looking at making similar changes to other blocks, for instance making the printer LEDs glow. Any feedback is welcome here!

---

As always, thank you for everyone who has contributed to this release!

**Comment by SirEdvin** [comment]
> Upgrades can now read and write NBT from the upgrade item when equipped. 

It is time to overhaul Turtlematic mod once more :)

**Comment by fatboychummy** [comment]
So turtles will now respect tool durability then? Nice, but also darn. I've gotten too used to turtles being infinite quarrying beasts. It'll take a bit to get used to that.

... I'll need to write some sort of mining program now that works with this new system. ~~Or maybe finish one of the many other ones I've started and never finished lol~~

**Reply by Wojbie** [reply]
No this is not active on default tools. If i understand it right this is something someone can add if they want, say they make custom turtle tools in their mod. See https://github.com/cc-tweaked/netherite-tools for example.

**Reply by SirEdvin** [reply]
Or you can overwrite default diamond tools to work with enchantments and/or use durability

**Reply by fatboychummy** [reply]
Ah, I misunderstood then. It just *unlocks* those features, not implements them.

**Reply by SquidDev** [reply]
Oh no, they are implemented, they're just not enabled by default.

**Reply by fatboychummy** [reply]
Goodness I need to get better at reading. I see all that in the original post now 🤦
